tsdb/errors/errors.go

```diff
@@ -17,21 +17,60 @@ package errors
 import (
 	"bytes"
 	"fmt"
+	"io"
 )
 
-// The MultiError type implements the error interface, and contains the
-// Errors used to construct it.
-type MultiError []error
+// multiError type allows to combine multiple errors into one.
+type multiError []error
 
-// Returns a concatenated string of the contained errors
-func (es MultiError) Error() string {
+// NewMulti returns multiError with provided errors added if not nil.
+func NewMulti(errs ...error) multiError { // nolint:golint
+	m := multiError{}
+	m.Add(errs...)
+	return m
+}
+
+// Add adds the error to the error list if it is not nil.
+// If the error has a nonNilMultiError error, errors from passed multi errors are combine in the
+// same level as existing errors.
+func (es *multiError) Add(errs ...error) {
+	for _, err := range errs {
+		if err == nil {
+			continue
+		}
+		if merr, ok := err.(nonNilMultiError); ok {
+			*es = append(*es, merr.errs...)
+			continue
+		}
+		*es = append(*es, err)
+	}
+}
+
+// Err returns the error list as an error or nil if it is empty.
+func (es multiError) Err() error {
+	if len(es) == 0 {
+		return nil
+	}
+	return nonNilMultiError{errs: es}
+}
+
+// nonNilMultiError implements the error interface, and it represents
+// multiError with at least one error inside it.
+// This type is needed to make sure that nil is returned when no error is combined in multiError for err != nil
+// check to work.
+type nonNilMultiError struct {
+	errs multiError
+}
+
+// Error returns a concatenated string of the contained errors.
+func (es nonNilMultiError) Error() string {
 	var buf bytes.Buffer
 
-	if len(es) > 1 {
-		fmt.Fprintf(&buf, "%d errors: ", len(es))
+	if len(es.errs) > 1 {
```

Comment:
if `nonMultiError` is only created when having at least 1 error, is this validation really necessary?

Reply:
Good point (: I guess not :+1:

Reply:
oops, I read if len > 0
probably need glasses lol